### PR TITLE
feat(pocket-ic): [NET-1725] Support synchronous call endpoint in pocket-ic.

### PR DESCRIPTION
**Pull request by DSharifi**
closes NET-1725

### CODE REVIEW

File: rs/pocket_ic_server/src/pocket_ic.rs [hunk 1]
```diff
@@ -1256,12 +1260,19 @@ impl Operation for StatusRequest {
...
+}
+
 pub struct CallRequest {
     pub effective_canister_id: CanisterId,
     pub bytes: Bytes,
     pub runtime: Arc<Runtime>,
+    pub version: CallRequestVersion,
 }
 
+const INGRESS_MESSAGE_CERTIFICATION_TIMEOUT_SECONDS: u64 = 10;
```
Comment (mraszyk) [inline rs/pocket_ic_server/src/pocket_ic.rs:1275]: Could you please move this constant to the top of the file and add a doc comment?

File: rs/pocket_ic_server/src/pocket_ic.rs [hunk 2]
```diff
@@ -83,6 +85,8 @@ include!(concat!(env!("OUT_DIR"), "/dashboard.rs"));
 /// Used for generating canister ID ranges that do not appear on mainnet.
 pub const MAXIMUM_NUMBER_OF_SUBNETS_ON_MAINNET: u64 = 1024;
 
+const EXECUTE_ROUND_INTERVAL: Duration = Duration::from_millis(50);
```
Comment (mraszyk) [inline rs/pocket_ic_server/src/pocket_ic.rs:88]: Could you please add a doc comment?

File: bazel/external_crates.bzl [hunk 3]
```diff
@@ -567,6 +569,12 @@ def external_crates_repository(name, cargo_lockfile, lockfile, sanitizers_enable
                     "pem",
                 ],
             ),
+            "ic-agent-call-v3": crate.spec(
+                package = "ic-agent",
+                git = "https://github.com/dsharifi/agent-rs",
```
Comment (basvandijk) [inline bazel/external_crates.bzl:574]: Why can't we use dfinity's repo here (https://github.com/dfinity/agent-rs)?

File: rs/pocket_ic_server/tests/test.rs [hunk 4]
```diff
@@ -685,3 +686,67 @@ fn canister_state_dir() {
     pic.update_call(nns_canister_id, Principal::anonymous(), "write", vec![])
         .unwrap();
 }
+
+/// Test that PocketIC can handle synchronous update calls, i.e. `/v3/.../call`.
```
Comment (mraszyk) [inline rs/pocket_ic_server/tests/test.rs:690]:
```suggestion
/// Test that PocketIC can handle synchronous update calls, i.e. `/api/v3/.../call`.
```

File: rs/pocket_ic_server/tests/test.rs [hunk 5]
```diff
@@ -685,3 +686,67 @@ fn canister_state_dir() {
...
+    let mut pic = PocketIcBuilder::new()
+        .with_nns_subnet()
+        .with_application_subnet()
+        .build();
+    let endpoint = pic.make_live(None);
+
+    // We define a "specified" canister ID that exists on the IC mainnet,
+    // but belongs to the canister ranges of no subnet on the PocketIC instance.
+    let specified_id = Principal::from_text("rimrc-piaaa-aaaao-aaljq-cai").unwrap();
+    assert!(pic.get_subnet(specified_id).is_none());
```
Comment (mraszyk) [inline rs/pocket_ic_server/tests/test.rs:705]:
```suggestion
```
not needed for this test

File: rs/pocket_ic_server/tests/test.rs [hunk 6]
```diff
@@ -685,3 +686,67 @@ fn canister_state_dir() {
...
+
+    // We create a canister with that specified canister ID: this should succeed
+    // and a new subnet should be created.
```
Comment (mraszyk) [inline rs/pocket_ic_server/tests/test.rs:708]:
```suggestion
```

outdated comment

File: rs/pocket_ic_server/tests/test.rs [hunk 7]
```diff
@@ -685,3 +686,67 @@ fn canister_state_dir() {
...
+
+        let arg = ProvisionalCreateCanisterWithCyclesArgs {
+            amount: None,
+            settings: None,
+            specified_id: None,
+            sender_canister_version: None,
+        };
+        let bytes = candid::Encode!(&arg).unwrap();
+
+        agent
```
Comment (mraszyk) [inline rs/pocket_ic_server/tests/test.rs:737]: Could you please add a comment that we're actually testing /api/v3/.../call here?

File: rs/pocket_ic_server/tests/test.rs [hunk 8]
```diff
@@ -685,3 +686,67 @@ fn canister_state_dir() {
...
+
+    // We create a canister with that specified canister ID: this should succeed
+    // and a new subnet should be created.
+    let rt = tokio::runtime::Builder::new_current_thread()
+        .enable_all()
+        .build()
+        .unwrap();
+
+    // retrieve the first canister ID on the application subnet
+    // which will be the effective and expected canister ID for canister creation
```
Comment (mraszyk) [inline rs/pocket_ic_server/tests/test.rs:715]:
```suggestion
    // which will be the effective canister ID for canister creation
```

File: rs/state_machine_tests/src/lib.rs [hunk 9]
```diff
@@ -621,7 +629,7 @@ pub struct StateMachine {
         tower::buffer::Buffer<IngressFilterService, (ProvisionalWhitelist, SignedIngressContent)>,
     payload_builder: Arc<RwLock<Option<PayloadBuilderImpl>>>,
     message_routing: SyncMessageRouting,
-    metrics_registry: MetricsRegistry,
+    pub metrics_registry: MetricsRegistry,
```
Comment (mraszyk) [inline rs/state_machine_tests/src/lib.rs:632]: Does it need to be public?

File: rs/pocket_ic_server/src/pocket_ic.rs [hunk 10]
```diff
@@ -1317,22 +1328,72 @@ impl Operation for CallRequest {
                 )
                 .build();
 
-                let svc = CallServiceV2::new_service(ingress_validator);
+                let subnet_clone = subnet.clone();
+
+                // Task that waits for call service to submit the ingress message, and
+                // forward it to the state machine. The task will automatically terminate
+                // once the call service sends a message, or if the call service is dropped.
```
Comment (mraszyk) [inline rs/pocket_ic_server/src/pocket_ic.rs:1335]:
```suggestion
                // Task that waits for call service to submit the ingress message, and
                // forwards it to the state machine. The task will automatically terminate
                // once it submits an ingress message received from the call service to the `StateMachine`, or if the call service is dropped (in which case `r.recv().await` fails with an error).
```

File: rs/pocket_ic_server/src/pocket_ic.rs [hunk 11]
```diff
@@ -1317,22 +1328,72 @@ impl Operation for CallRequest {
...
+                    let cancellation_token_clone = cancellation_token.clone();
+
+                    s.spawn(move || {
+                        while !&cancellation_token_clone.is_cancelled() {
+                            for subnet in pic.subnets.read().unwrap().values() {
+                                subnet.execute_round();
+                            }
+                            std::thread::sleep(EXECUTE_ROUND_INTERVAL);
+                        }
+                    });
```
Comment (mraszyk) [inline rs/pocket_ic_server/src/pocket_ic.rs:1391]: Let's only spawn this task for V3, please.

File: rs/pocket_ic_server/src/pocket_ic.rs (hunk 10 above)
Comment (mraszyk) [reply rs/pocket_ic_server/src/pocket_ic.rs:1335]: please double check my suggestion before applying it

File: bazel/external_crates.bzl (hunk 3 above)
Comment (DSharifi) [reply bazel/external_crates.bzl:574]: We don't want to merge the changes to agent-rs yet, and I do not have write access to create branches in the agent-rs repo.

File: rs/state_machine_tests/src/lib.rs (hunk 9 above)
Comment (DSharifi) [reply rs/state_machine_tests/src/lib.rs:632]: No, leftover

File: rs/pocket_ic_server/src/pocket_ic.rs (hunk 10 above)
Comment (DSharifi) [reply rs/pocket_ic_server/src/pocket_ic.rs]: Changed your suggestion  slightly

**Review by mraszyk — APPROVED** [review]

File: bazel/external_crates.bzl (hunk 3 above)
Comment (rumenov) [reply bazel/external_crates.bzl:574]: Also we do have precedents, e.g. the  git = "https://github.com/blind-oracle/cloudflare-rs.git",

Comment (basvandijk) [reply bazel/external_crates.bzl:574]: > I do not have write access to create branches in the agent-rs repo.

I just gave the [core](https://github.com/orgs/dfinity/teams/core) team Write access to dfinity/agent-rs. Everybody should be able to push code to that repo. We have branch protection in place to protect the `main` branch.

File: bazel/external_crates.bzl [hunk 12]
```diff
@@ -567,6 +569,12 @@ def external_crates_repository(name, cargo_lockfile, lockfile, sanitizers_enable
                     "pem",
                 ],
             ),
+            "ic-agent-call-v3": crate.spec(
```
Comment (rumenov) [inline bazel/external_crates.bzl:572]: Please add a todo with. a jira ticket attached that we should delete this once the feature is done

**Review by schneiderstefan — APPROVED** [review]

File: bazel/external_crates.bzl (hunk 3 above)
Comment (DSharifi) [reply bazel/external_crates.bzl]: Changed it to use a revision of dfinity's repo

**Review by basvandijk — APPROVED** [review]

**Review by mraszyk — APPROVED** [review]

File: rs/pocket_ic_server/src/pocket_ic.rs [hunk 13]
```diff
@@ -1266,9 +1272,16 @@ impl Operation for StatusRequest {
...
 
+pub enum CallRequestVersion {
+    V2,
+    V3,
+}
+
 pub struct CallRequest {
     pub effective_canister_id: CanisterId,
     pub bytes: Bytes,
+    pub runtime: Arc<Runtime>,
```
Comment (mraszyk) [inline rs/pocket_ic_server/src/pocket_ic.rs:1283]: This runtime should not be needed; `pic.runtime` can be used instead.

File: rs/pocket_ic_server/src/pocket_ic.rs [hunk 14]
```diff
@@ -1326,22 +1339,77 @@ impl Operation for CallRequest {
...
                         PrincipalId(self.effective_canister_id.get().into())
                     ))
                     .body(self.bytes.clone().into())
                     .unwrap();
-                let resp = pic.runtime.block_on(svc.oneshot(request)).unwrap();
 
-                if let Ok(UnvalidatedArtifactMutation::Insert((msg, _node_id))) = r.try_recv() {
-                    subnet.push_signed_ingress(msg);
-                }
+                // but fine for now to unblock agent-rs.
```
Comment (mraszyk) [inline rs/pocket_ic_server/src/pocket_ic.rs:1391]: This seems to be missing the beginning of the sentence.

**Comment by fxgst** [comment]
Do we need to update the CHANGELOG?

**Comment by DSharifi** [comment]
> Do we need to update the CHANGELOG?

Added it.

**Review by rumenov — APPROVED** [review]

**Pull request merged**